### ISSUE TITLE
After a migration that added a GIN, all subsequent migrations drop and recreate this index

### ISSUE THREAD

**Issue by AndrewCathcart** [prisma/prisma#16275]
### Bug description

We recently added a Generalized Inverted Index to one of our tables. 

```sql
CREATE EXTENSION IF NOT EXISTS pg_trgm;
CREATE EXTENSION IF NOT EXISTS btree_gin;

-- CreateIndex
CREATE INDEX "workspace_folder_workspaceId_status_name_idx" ON "workspace_folder" USING GIN ("workspaceId", "status" , "name" gin_trgm_ops);
```

All subsequent migrations now drop this index and recreate it, for example;

```sql
-- DropIndex
DROP INDEX "workspace_folder_workspaceId_status_name_idx";

-- CreateIndex
CREATE INDEX "workspace_invite_wid_index" ON "WorkspaceInvite"("workspaceId");

-- CreateIndex
CREATE INDEX "workspace_folder_workspaceId_status_name_idx" ON "workspace_folder" USING GIN ("workspaceId" , "status" , "name" gin_trgm_ops);
```

### Expected behavior

It should not drop and recreate the GIN on subsequent migrations

### Prisma information

```prisma
//...
model WorkspaceInvite {
  userId      String                @db.Uuid
  workspaceId String                @db.Uuid
  role        WorkspaceRole         @default(editor)
  status      WorkspaceInviteStatus
  createdAt   DateTime              @default(now())
  updatedAt   DateTime              @default(now()) @updatedAt
  invitedBy   String?               @db.Uuid
  workspace   Workspace             @relation(fields: [workspaceId], references: [id])

  @@id([userId, workspaceId])
  @@unique([userId, workspaceId])
  @@index([workspaceId])
}

model WorkspaceFolder {
  id                String                  @id @default(uuid()) @db.Uuid
  name              String
  workspaceId       String                  @db.Uuid
  parentId          String?                 @db.Uuid
  createdAt         DateTime                @default(now())
  status            WorkspaceFolderStatus   @default(active)
  updatedAt         DateTime                @default(now()) @updatedAt
  category          WorkspaceFolderCategory @default(default)
  parent            WorkspaceFolder?        @relation("WorkspaceFolderToWorkspaceFolder", fields: [parentId], references: [id], onDelete: Cascade)
  workspace         Workspace               @relation(fields: [workspaceId], references: [id])
  WorkspaceProjects WorkspaceProject[]
  WorkspaceFolder   WorkspaceFolder[]       @relation("WorkspaceFolderToWorkspaceFolder")
  WorkspaceSettings WorkspaceSettings[]

  @@index([workspaceId], map: "workspace_folder_wid_index")
  @@index([parentId], map: "parent_id_folder_index")
  @@index([workspaceId, parentId], map: "workspace_id_folder_index")
  @@index([workspaceId, parentId, status], map: "workspace_id_folder_index_status")
  @@index([workspaceId, parentId, status, category])
  @@index([workspaceId(ops: raw("")), status(ops: raw("")), name(ops: raw("gin_trgm_ops"))], type: Gin)
  @@map("workspace_folder")
}
// ...
```

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: v18.12.0

### Prisma Version

```
8.19.2
```

**Comment by SevInf** [comment]
Hi @AndrewCathcart !
Which version of Prisma are you using? You've reported `8.19.2`, but this version does not exist.

**Comment by AndrewCathcart** [comment]
Hi @SevInf apologies, copy pasted the wrong value. Currently using `4.3.1`

**Comment by AndrewCathcart** [comment]
Let me know if you need any more information to reproduce this

**Comment by janpio** [comment]
Is the generated SQL 100% identical to what you initially created? If so, sounds like a bug in the diffing of the `raw` ops logic, making Migrate think that it is different each time and needs to be  
Does the `workspace_invite_wid_index` index make sense to you?

**Comment by AndrewCathcart** [comment]
Yes sorry I should have been a bit clearer - the workspace_invite_wid_index is the index I added, but I was expecting that to be the only line in the migration.

To be 100% clear, this is the original SQL in the first migration where the GIN was added;

```sql
CREATE EXTENSION IF NOT EXISTS pg_trgm;
CREATE EXTENSION IF NOT EXISTS btree_gin;

-- CreateIndex
CREATE INDEX "workspace_folder_workspaceId_status_name_idx" ON "workspace_folder" USING GIN ("workspaceId", "status" , "name" gin_trgm_ops);
```

And here is what gets generated in every migration now, for example running migrate dev on a fresh database;

```sql
-- DropIndex
DROP INDEX "workspace_folder_workspaceId_status_name_idx";

-- CreateIndex
CREATE INDEX "workspace_folder_workspaceId_status_name_idx" ON "workspace_folder" USING GIN ("workspaceId" , "status" , "name" gin_trgm_ops);
```

**Comment by uhyo** [comment]
:+1:, I just got blocked by this today :sob:

A bit of additional information: running `prisma db pull` removes `ops: raw("")` from the schema.

Before `prisma db pull`

```prisma
@@index([companyId(ops: raw("")), text(ops: raw(""))], type: Gin)
```

Aftter `prisma db pull`

```prisma
@@index([companyId, text], type: Gin)
```

**Comment by uhyo** [comment]
So I created a pull request. Hope you can check this :sob:

https://github.com/prisma/prisma-engines/pull/3625

**Comment by Druue** [comment]
Managed to get a minimal repro up for this in 4.9.0: [druue/prisma-debug](https://github.com/Druue/prisma-debug/tree/repro/migrate-gin)

**Comment by janpio** [comment]
Thanks for your PR @uhyo. This triggered me to reproduce this again based on the reproduction that @Druue posted above, as I really wanted to understand what is going here to be able to decide if your PR is doing the _right_ thing and does not just hide the problem.

Soooo:

1. Prisma does not know about the (default) ops classes that become available for fields of type `String` and `@db.Uuid` when you install both extensions `pg_trgm` and `btree_gin`. That is why the schema becomes invalid when you add the fields to the Gin index with an ops class.
2. By adding `ops: raw("")` you work around that - as the validation logic is currently configured to just accept whatever string in there as a valid ops class that it tries to generate SQL for. At the same time you modify the `migration.sql` to just add the two extensions and use the default ops class and that works fine.
3. Now when you run `migrate dev` the next time, your migration history will be applied to your shadow database (which will work fine), and then it will "introspect" the result and be told that the field `workspaceId` has an ops class `uuid_ops` and that this is the default ops class for that field in the database. As that does not match the current schema file (which has `ops: raw("")` - which is not the same as "nothing") this is raised in our diffing and it decides it needs to create a migration where the ops class is an empty string. This creates the endless cycle described here in this issue.

Now your suggestion @uhyo in the PR is to treat `raw("")` as "nothing" basically. That will lead the diffing to skip this, and no new migration to be generated.

I am not sure that is the right way to go here. `ops: raw("")` really is not the same as no ops class.
Based on that I would suggest that we do _not_ accept your PR.

---

What can you do instead?

The real solution from Prisma of course is to understand that the extensions are installed via [our `extensions` feature (currently  `postgresqlExtensions` preview feature)](https://www.prisma.io/docs/concepts/components/prisma-schema/postgresql-extensions) - but that will take a while until we get to that (but I will create an explicit issue for that so we can better track that). So we need something better for you in the meantime.

Have you considered to use the actual ops class that is being added by the extension in the meantime?
```
    @@index([workspaceId(ops: raw("uuid_ops")), status(ops: raw("uuid_ops")), name(ops: raw("gin_trgm_ops"))], type: Gin)
```
In my testing this seems to make `migrate dev` happy and does not create endless migrations any more. `db pull` will still remove the explicit ops class though right now, but that should not be a big problem for you.

(If that works for you, I think we might actually invalidate `raw("")` as an ops class soon. This should never be necessary and you saw the problems it leads to. We better want to hear if this causes any trouble - like disabling your workaround - and then figure out how to really avoid these problems)

**Comment by uhyo** [comment]
@janpio Thank you for your investigation. I understand that it'd be better to explicitly write down an explicit operator class.

Actually that's why I also included treatment of cases like `raw("text_ops")` in that PR. This one still leads to the infinite cycle.

I'm willing to remove special-casing of `raw("")` but it'd be nice of you could accept the remaining.

**Comment by janpio** [comment]
Ok, I missed that one in the PR as the reproduction and earlier comments mostly talked about the `raw("")` case. I created a separate issues https://github.com/prisma/prisma/issues/17518 for the `text_ops` case which I can also reproduce, but has to have a slightly different root cause. Please jump over there and double check if everything I wrote is correct and we can continue the discussion there. Thanks.

**Comment by AndrewCathcart** [comment]
Using uuid_ops and enum_ops seems to be acceptable for now

**Comment by AndrewCathcart** [comment]
@janpio Given this hasn't received any feedback / movement in a while and that using uuid_ops and enum_ops seemed to be fine as a workaround, would you like me to close this issue? Alternatively if you have a little bit of direction I could try and fix this myself If I get some free time.
